### PR TITLE
X360 controller support for Windows + Rumble

### PR DESCRIPTION
**Pull request by KrossX**
Some basic controller input for Windows, but rumble required a PuruPuru Pack so added that too.

Quite basic, no error check neither fancy continuous rumble mode but it works. Rez feels nice~

### CODE REVIEW

**Comment by hooby3dfx** [comment]
looks good! was super happy to see this. i will try and add in some android handling.

**Comment by hooby3dfx** [comment]
once merged this should close https://github.com/reicast/reicast-emulator/issues/340

**Comment by baka0815** [comment]
Looks good, but shouldn't this be two separate pull requests?

One to add the purupuru support and one for the XBox gamepad?

**Comment by hooby3dfx** [comment]
That's one of those cases where you gotta ask 'does it really matter'?

**Comment by KrossX** [comment]
@baka0815 Yes, it should've been different pulls. The XInput code is easy to separate if needed, but it contains the rumble implementation which I thought to be important.

**Comment by baka0815** [comment]
Ok, yeah, I take the nitpicking award. :-)

**Comment by skmp** [comment]
Oi, @KrossX, great to see you around :D. I'll take a look later today and possibly merge ~

@baka0815 ideally yes, but the changes look small enough anyway. From my POV getting code in is more important than too strict PR rules

**Comment by Shoegzer** [comment]
@KrossX the purupuru stuff is really nice to see - thanks!  

@Holzhaus any chance you'd consider porting this to Linux to add to all your great input work there?

**Comment by KrossX** [comment]
It's razy! Hello~, I mean @skmp ... skimpy?

**Comment by Holzhaus** [comment]
@Shoegzer Although I'm quite busy atm,  I'd love to add rumble support on Linux. However, I need some more info on the u32 value. What exactly is described there?

**Comment by KrossX** [comment]
@Holzhaus, check the [RumbleConfig](https://github.com/KrossX/Pokopom/blob/ec9f21690f7714dd395e1ee559706e3ca05bbbc8/Pokopom/nullDC_Devices.h#L106-L106) struct.

I dunno how versatile can be a mobile rumble, so maybe all that matters is the intensity.

**Comment by Shoegzer** [comment]
@Holzhaus thanks so much for looking into it, whenever you have time - I'm patient!  If/when you get to a point that you need help testing, just let me know.  :)

**Comment by Holzhaus** [comment]
Ok, here's my initial attempt: https://github.com/Holzhaus/reicast-emulator/commit/cc5719e84524e1bea7940d7abff3f24c76157eb0

I just used the same code that KrossX wrote for the windows version, had a look at [https://msdn.microsoft.com/en-us/library/windows/desktop/microsoft.directx_sdk.reference.xinput_vibration(v=vs.85).aspx](docs for the XINPUT_VIBRATION structure) and came up with this probably super inaccurate analogy.

| Motor type | Windows (XINPUT)   | Linux (Force FB)   |
| ---------- | ------------------ | ------------------ |
| low freq   | `wLeftMotorSpeed`  | `strong_magnitude` |
| high freq  | `wRightMotorSpeed` | `weak_magnitude`   |

This enabled rumble support in linux, but I can't believe it's actually works as intended. While Soul Calibur seems to work okay, my controller is almost rumbling the whole time when playing Crazy taxi, which is super annoying.

@Shoegzer For testing, you need to undo commit https://github.com/KrossX/reicast-emulator/commit/c42986e63f167b73493209ab45b306522671fa86.

**Comment by LoungeKatt** [comment]
Adding support for Android and iOS devices directly would be fun, but probably not while it's going to turn it into a pager on crack.

**Comment by KrossX** [comment]
@Holzhaus Yup, that's about it. It's quite the strong rumble and Crazy Taxy rumbles on everything and even uses the direction change on crashes. A simple intensity setting would do the trick.

**Comment by Shoegzer** [comment]
@Holzhaus Wow, that was fast - thanks so much!  @KrossX thanks to you too of course.  :)

Tested and working very well here.  Yes, Crazy Taxi rumble is crazy but I guess it's normal.  Other games behave better though - in particular: Daytona USA feels right, hitting walls has an entirely different feeling than driving over grass, as to be expected.  Also, an emu.cfg option and hotkey setting to add/remove the Jump (Purupuru) Pack would be nice.

Aside from that, can you issue a PR to main?  And... let's get Ready 2 Rumble!  (couldn't resist).  ;-)

**Comment by baka0815** [comment]
Later this day I will try to check Crazy Taxi on my Dreamcast to see if it also rumbles like crazy.

**Comment by baka0815** [comment]
Ok, Crazy Taxi rumbles the whole time even on the real Dreamcast but very lightly (I assume it should be the motor running). However I can't run your fork @Holzhaus.

While compiling I get lots of `cc1: warning: command line option ‘-fno-rtti’ is valid for C++/ObjC++ but not for C` messages I don't get using master.

It starts and then dies while initializing the renderer:

```
# ./reicast.elf /.../Crazy\ Taxi.cdi 
Home dir is: /home/baka/.reicast/
Linux paging: 00000000 00000000 00000FFF
evdev open: No such file or directory
SDK: Found 'Logitech Logitech RumblePad 2 USB' joystick with 6 axis and 12 buttons
MAP 00800000 w/ 25165824
MAP 20000000 w/ 25165824
MAP 04000000 w/ 16777216
MAP 06000000 w/ 16777216
MAP 0C000000 w/ 0
A8
vmem reserve: base: D6B50000, aram: f6b50000, vram: DAB50000, ram: E2B50000
Resetting mem
Mem alloc successful!Using '/.../Crazy Taxi.cdi' as cd image

Malformed entry on config - ignoring @ 47(ElfFile)
libwebsockets test server - (C) Copyright 2010-2013 Andy Green <andy@warmcat.com> - licensed under LGPL2.1
Initial logging level 7
Library version: 1.3 c11b847
IPV6 not compiled in
libev support not compiled in
 static allocation: 4452 + (12 x 1024 fds) = 16740 bytes
 canonical_hostname = baka
 per-conn mem: 116 + 1554 headers + protocol rx buf
 Listening on port 5678
Loaded /home/baka/.reicast/data/dc_boot.bin as bootrom

Loaded /home/baka/.reicast/data/dc_flash.bin as nvram

Using Interpreter

Session 1 has 1 track(s)
Saving  Track:  1  Type: Audio/2352  Pregap: 150  Size: 302     LBA: 0       
Track position: 352800

Session 2 has 1 track(s)
Saving  Track:  2  Type: Mode2/2336  Pregap: 150  Size: 309642  LBA: 11702   
Track position: 1413504
gdrom: Opened image "/.../Crazy Taxi.cdi"
Initialising empty vmu...
Sh4 Reset
Sh4: i-cache invalidation A0000020
GL context version: 3.0
Fatal error : rend->init() failed

 in rend_thread -> ../../core/hw/pvr/Renderer_if.cpp : 245 

Trace/Breakpoint ausgelöst
```

So I checked out @KrossX fork and merged your commit manually and it looks like the behaviour is the same as the real DC, smooth rumbling.
But when I crush into other cars the rumbling doesn't increase.

That's because you got the order wrong. It has to be the following (weak first)
```
void input_evdev_rumble(EvdevController* controller, u16 pow_weak, u16 pow_strong)
```

**Comment by baka0815** [comment]
@Holzhaus 
Change the code in `core/linux-dist/main.cpp` to
```
u16 pow_weak = (u16)(65535 * pow_l);
u16 pow_strong = (u16)(65535 * pow_r);
```
(see my comment in your commit).
This won't confuse anyone like my ordering suggestion from above.

**Comment by KrossX** [comment]
That is odd. With the X360 controller, the left and right motors were indeed the strong and weak ff_effect setting. Maybe it depends on the driver?

**Comment by baka0815** [comment]
I'm talking about the evdev implementation, tried using a Logitech RumblePad 2.

Unfortunately I don't own a XBox controller.

**Comment by baka0815** [comment]
I uploaded a short video of Crazy Taxi: https://youtu.be/AMRt7f8XcrU (the fun starts at 1:00).

You can see that, in fact, it's rumbling while colliding with other cars. Please excuse the crappy quality and  perspective - it's not so easy driving while filming with your mobile phone. ;-)

The version I'm using in the video uses "my" reversed version of strong and weak.

**Comment by KrossX** [comment]
Is that about the crashing rumble? The effect I mentioned is when the rumble motor of the PuruPuru changes direction, which supposedly gives a strong effect when it happens. That doesn't mean there wouldn't be rumble, that's why I didn't bother. =P

**Comment by Holzhaus** [comment]
> I'm talking about the evdev implementation, tried using a Logitech RumblePad 2.
> ```C
> u16 pow_weak = (u16)(65535 * pow_l);
> u16 pow_strong = (u16)(65535 * pow_r);
> ```

Well, this implies that `pow_l` is used for weak rumble and `pow_r` is used the strong rumble. The XINPUT (Windows) by KrossX actually does the opposite. It doesn't make sense to use a value for weak rumbling on one platform and strong rumbling on another. 

@KrossX: Maybe your implementation accidentally swapped the values for left and right?

Or the driver for the Logitech RumblePad 2 is buggy on linux, i.e. the strong/weak motor is reversed?

**Comment by KrossX** [comment]
I made some evdev rumble work for PS1/2 emulation. You chart was correct, at least for the x360 controller. Using evdev, left motor corresponded with the strong setting and right motor with the weak. Left motor is the big motor, used for low_freq on PuruPuru; right motor is the small motor, used for high_freq on PuruPuru. 

It could be a controller difference, maybe even both motors are the same or swapped. Could be a driver difference, as I don't even know if there's a standard for that. Or I could've bugged the frequency part of the code, which is based on optional code by Falcon.

I'll recheck the code just in case.

**Comment by baka0815** [comment]
I don't know if I will have time to test your changes this year, but I will notify you when I did.

**Comment by baka0815** [comment]
@KrossX If I revert my changes and apply your latest patch to ```/core/linux-dist/main.cpp``` than the rumble works as expected. However the rumble when crashing into stuff is a lot weaker than with the reversed version.
So the changes should be merged from my POV!

**Comment by Shoegzer** [comment]
@KrossX Out of curiosity, where does this stand now?  I'm guessing it's ready to merge but didn't want to assume in case you need more help testing anything.

**Comment by KrossX** [comment]
Yes, I would consider it ready indeed.

**Comment by Shoegzer** [comment]
Ah ok, thanks, i figured such.  @skmp or @Holzhaus can you merge this when you a minute please?

**Comment by baka0815** [comment]
With the additions from @Holzhaus and the latest change from @KrossX which also needs to be done for ```/core/linux-dist/main.cpp``` than the changes this should be merged upstream!

**Comment by Holzhaus** [comment]
I'm currently very busy, and I'm a bit hesitant to merge it without retesting it properly. Maybe @skmp can look into this? If not, you can just compile reicast with these patches merged manually.

**Comment by KrossX** [comment]
That reminds me, it still needs a setting to plug the rumble device. That's why I unplugged it, so as to make it an opt-in feature.

I have no clue how settings are handled though.

**Comment by baka0815** [comment]
I don't think that there is a possibility for this yet, but @Holzhaus works on a GUI which would handle that AFAIK.
The branch should be merged nonetheless because the addition is worth it.

@Holzhaus I tested your branch with the latest addition from @KrossX and it behaves like the real DC controller on my end. Do you need anything specific to remove your worries?

**Comment by Holzhaus** [comment]
Did you test both Windows and Linux?

File: core/windows/winmain.cpp [hunk 1]
```diff
@@ -245,7 +249,66 @@ void UpdateInputState(u32 port)
...
+			rt[port] |= state.Gamepad.bRightTrigger;
+
+			joyx[port] |=  state.Gamepad.sThumbLX / 257;
+			joyy[port] |= -state.Gamepad.sThumbLY / 257;
+		}
+	}
+
+void UpdateVibration(u32 port, u32 value)
+{
+		const u8 freq_l = 0x16;
```
Comment (Holzhaus) [inline core/windows/winmain.cpp]: These two lines are obsolete, aren't they?

Comment (KrossX) [inline core/windows/winmain.cpp]: freq_l and freq_h ones? Yes, I suppose those don't even work as reference anymore.

**Comment by baka0815** [comment]
I tested only your evdev-Patches on Linux.

**Comment by Shoegzer** [comment]
@Holzhaus Linux side also tests successfully on this end at least, and from what @KrossX said it sounds like the Windows side is good too(?) though I haven't tried myself.  This leaves out the ability to plug in the rumble pack, which would still be a nice feature but not strictly necessary.  Anyway, whenever you have time.. no rush I guess.

**Comment by Zetchzie** [comment]
when will this be put into the nightlies, and if it is where do i get the config files to use xinput? This cant come soon enough >"<

**Comment by baka0815** [comment]
What's missing for this to be merged? Time or is it something different?

**Comment by Holzhaus** [comment]
Can't speak for the ozher drvs, but I'm out of time at the moment. This will hopefully change end of March.

**Comment by LoungeKatt** [comment]
Same here. Scheduled everything for "early next year" and this is it.

**Comment by baka0815** [comment]
Ok, Reicast not dead - some pull requests got merged.

What can I do for this to get merged? Would it help if I took the work from @KrossX and @Holzhaus in my own branch and create a new pull request with all the settings including Windows and Linux?

I don't see any problems with this work, the rumble works as expected and the possibility for configuration can follow later.

**Comment by baka0815** [comment]
Again, I'm still willing to help get this merged.

**Comment by skmp** [comment]
@KrossX @LoungeKatt @Holzhaus Is this in a mergable state? I'll test over the weekend and merge if no one disagrees. Gotta start cleaning up the pending PRs!

**Comment by KrossX** [comment]
I don't remember, I can see that at least the rumble is unplugged so no change there but there's a dependency on XInput9_1_0 on Windows.

Being able to choose between VMU or Rumble requires a setting or something. The dll dependency might not be a problem as it's included in DirectX runtimes.

**Comment by Holzhaus** [comment]
Can be merged since it doesn't break something, but someone should add an config flag to enable rumble support (it's disabled now).

**Comment by Zetchzie** [comment]
yay finally joypad support

**Comment by baka0815** [comment]
You should merge the linux rumble support PR from @Holzhaus (#964) because he included his fixes for the evdev implementation there.
This one is - from the Linux pov - imcomplete.

**Pull request merged**